File: 2025_Oct/Kyle/2025-11-02-3Sum.py
class Solution(object):
    def threeSum(self, nums):
        nums.sort()
        output = []

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            target = -nums[i]
            left = i + 1
            right = len(nums) - 1

            while left < right:
                if nums[left] + nums[right] == target:
                    output.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while left < right and nums[left] == nums[left - 1]:
                        left += 1
                    while left < right and nums[right] == nums[right + 1]:
                        right -= 1
                elif nums[left] + nums[right] < target:
                    left += 1
                else:
                    right -= 1

        return output

File: 2025_Oct/Jayden/2025-11-02-3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        /**
            Time complexity: O(n^2)
            Space complexity: O(1) or O(k) where k is nuber of triplets
         */
        Arrays.sort(nums);

        List<List<Integer>> result = new ArrayList<>();

        for (int pivot = 0; pivot < nums.length - 2; pivot++) {

            // skip duplicates
            if (pivot > 0 && nums[pivot] == nums[pivot - 1])
                continue;

            int left = pivot + 1;
            int right = nums.length - 1;

            while (left < right) {
                int sum = nums[pivot] + nums[left] + nums[right];

                if (sum == 0) {
                    result.add(Arrays.asList(nums[pivot], nums[left], nums[right]));

                    // skip duplicates
                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;

                    left++;
                    right--;

                // if sum is < 0 increase left pointer by 1 otherwise, decrease the right pointer
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }

        }

        return result;
    }
}